src/dialects/oracle/query-generator.js

```diff
@@ -0,0 +1,1443 @@
+
+'use strict';
+
+/* jshint -W110 */
+const Utils = require('../../utils');
+const DataTypes = require('../../data-types');
+const AbstractQueryGenerator = require('../abstract/query-generator');
+const _ = require('lodash');
+const util = require('util');
+const Transaction = require('../../transaction');
+
+/**
+ * list of reserved words in Oracle DB 21c
+ * source: https://docs.oracle.com/en/cloud/saas/taleo-enterprise/21d/otccu/r-reservedwords.html#id08ATA0RF05Z
+ *
+ * @private
+ */
+const ORACLE_RESERVED_WORDS = ['ACCESS', 'ADD', 'ALL', 'ALTER', 'AND', 'ANY', 'ARRAYLEN', 'AS', 'ASC', 'AUDIT', 'BETWEEN', 'BY', 'CHAR', 'CHECK', 'CLUSTER', 'COLUMN', 'COMMENT', 'COMPRESS', 'CONNECT', 'CREATE', 'CURRENT', 'DATE', 'DECIMAL', 'DEFAULT', 'DELETE', 'DESC', 'DISTINCT', 'DROP', 'ELSE', 'EXCLUSIVE', 'EXISTS', 'FILE', 'FLOAT', 'FOR', 'FROM', 'GRANT', 'GROUP', 'HAVING', 'IDENTIFIED', 'IMMEDIATE', 'IN', 'INCREMENT', 'INDEX', 'INITIAL', 'INSERT', 'INTEGER', 'INTERSECT', 'INTO', 'IS', 'LEVEL', 'LIKE', 'LOCK', 'LONG', 'MAXEXTENTS', 'MINUS', 'MODE', 'MODIFY', 'NOAUDIT', 'NOCOMPRESS', 'NOT', 'NOTFOUND', 'NOWAIT', 'NULL', 'NUMBER', 'OF', 'OFFLINE', 'ON', 'ONLINE', 'OPTION', 'OR', 'ORDER', 'PCTFREE', 'PRIOR', 'PRIVILEGES', 'PUBLIC', 'RAW', 'RENAME', 'RESOURCE', 'REVOKE', 'ROW', 'ROWID', 'ROWLABEL', 'ROWNUM', 'ROWS', 'SELECT', 'SESSION', 'SET', 'SHARE', 'SIZE', 'SMALLINT', 'SQLBUF', 'START', 'SUCCESSFUL', 'SYNONYM', 'SYSDATE', 'TABLE', 'THEN', 'TO', 'TRIGGER', 'UID', 'UNION', 'UNIQUE', 'UPDATE', 'USER', 'VALIDATE', 'VALUES', 'VARCHAR', 'VARCHAR2', 'VIEW', 'WHENEVER', 'WHERE', 'WITH'];
+const JSON_FUNCTION_REGEX = /^\s*((?:[a-z]+_){0,2}jsonb?(?:_[a-z]+){0,2})\([^)]*\)/i;
+const JSON_OPERATOR_REGEX = /^\s*(->>?|@>|<@|\?[|&]?|\|{2}|#-)/i;
+const TOKEN_CAPTURE_REGEX = /^\s*((?:([`"'])(?:(?!\2).|\2{2})*\2)|[\w\d\s]+|[().,;+-])/i;
+
+class OracleQueryGenerator extends AbstractQueryGenerator {
+  constructor(options) {
+    super(options);
+  }
+
+  /**
+   * Returns the value as it is stored in the Oracle DB
+   *
+   * @param {string} value
+   */
+  getCatalogName(value) {
+    if (value) {
+      if (this.options.quoteIdentifiers === false) {
+        const quotedValue = this.quoteIdentifier(value);
+        if (quotedValue === value) {
+          value = value.toUpperCase();
+        }
+      }
+    }
+    return value;
+  }
+
+  /**
+   * Returns the tableName and schemaName as it is stored the Oracle DB
+   *
+   * @param {object|string} table
+   */
+  getSchemaNameAndTableName(table) {
+    const tableName = this.getCatalogName(table.tableName || table);
+    const schemaName = this.getCatalogName(table.schema);
+    return [tableName, schemaName];
+  }
+
+  createSchema(schema) {
+    const quotedSchema = this.quoteTable(schema);
+    const schemaName = this.getCatalogName(schema);
+    return [
+      'DECLARE',
+      '  V_COUNT INTEGER;',
+      '  V_CURSOR_NAME INTEGER;',
+      '  V_RET INTEGER;',
+      'BEGIN',
```

Comment:
Why not save a round-trip and simply execute the CREATE USER, then capture and ignore any 'already exists' exception?

Reply:
used a flag to check for user existence and an exception handler. removed select query .